File: Classes/objc/GBSlideOutToUnlockView.m
//
//  GBSlideOutToUnlockView.m
//  GBSlideOutToUnlockView
//
//  Created by Gustavo Barbosa on 4/16/14.
//  Copyright (c) 2014 Gustavo Barbosa. All rights reserved.
//

#import "GBSlideOutToUnlockView.h"

#import <QuartzCore/QuartzCore.h>

CGPoint p_fixedCenter(CGRect frame) {
    CGFloat _fixedX = (CGRectGetMaxX(frame) - CGRectGetMinX(frame)) / 2.0;
    CGFloat _fixedY = (CGRectGetMaxY(frame) - CGRectGetMinY(frame)) / 2.0;
    return CGPointMake(_fixedX, _fixedY);
}

static CGFloat const kDefaultInnerCircleRadius = 25.0f;

@interface GBSlideOutToUnlockView ()
{
    BOOL _unlockOnRelease;
    CGFloat _distanceMoved;
}

@property (nonatomic, strong) UIButton *dragButton;
@property (nonatomic, strong) UIView *animationDotView;

@end


@implementation GBSlideOutToUnlockView

@synthesize draggableButtonBackgroundColor = _draggableButtonBackgroundColor;

#pragma mark - Ctors

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self setupViews];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
#if !TARGET_INTERFACE_BUILDER
        [self setupViews];
#endif
    }
    return self;
}

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self setupViews];
    }
    return self;
}

#pragma mark - Setup

- (void)setupViews
{
    self.backgroundColor = [UIColor clearColor];
    [self addRedeemImageAtCenter];
    [self addPanGestureRecognizerToImage];
}

- (void)prepareForInterfaceBuilder
{
    [self setupViews];
}

#pragma mark - Acessors

- (CGFloat)innerCircleRadius
{
    return _innerCircleRadius ?: kDefaultInnerCircleRadius;
}

- (CGFloat)outerCircleRadius
{
    if (_outerCircleRadius)
        return _outerCircleRadius;

    CGFloat horizontalSize = CGRectGetWidth(self.bounds);
    CGFloat verticalSize = CGRectGetHeight(self.bounds);

    return MIN(horizontalSize, verticalSize) / 2.0f;
}

- (UIColor *)innerCircleColor
{
    return _innerCircleColor ?: self.tintColor;
}

- (UIColor *)outerCircleColor
{
    return _outerCircleColor ?: self.tintColor;
}

- (UIColor *)draggableButtonBackgroundColor
{
    return _draggableButtonBackgroundColor ?: self.tintColor;
}

- (void)setDraggableButtonBackgroundColor:(UIColor *)draggableButtonBackgroundColor
{
    _draggableButtonBackgroundColor = draggableButtonBackgroundColor;
    _dragButton.backgroundColor = _draggableButtonBackgroundColor;
}

- (UIColor *)draggableImageTintColor
{
    return _draggableImageTintColor ?: [UIColor whiteColor];
}

#pragma mark - Drawing

- (void)drawRect:(CGRect)rect
{
    [self drawCircleWithRadius:self.innerCircleRadius color:self.innerCircleColor];
    [self drawCircleWithRadius:self.outerCircleRadius color:self.outerCircleColor];
}

- (void)drawCircleWithRadius:(CGFloat)radius color:(UIColor *)color
{
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetLineWidth(context, 1.0);
    CGContextSetStrokeColorWithColor(context, color.CGColor);

    CGRect circle = CGRectMake(p_fixedCenter(self.frame).x - radius,
                               p_fixedCenter(self.frame).y - radius,
                               2 * radius,
                               2 * radius);

    CGContextAddEllipseInRect(context, circle);
    CGContextStrokePath(context);
}

- (void)addRedeemImageAtCenter
{
    _dragButton = [UIButton buttonWithType:UIButtonTypeCustom];
    _dragButton.frame = CGRectMake(0, 0, 2*self.innerCircleRadius, 2*self.innerCircleRadius);
    _dragButton.backgroundColor = self.draggableButtonBackgroundColor;
    _dragButton.tintColor = self.draggableImageTintColor;
    _dragButton.layer.cornerRadius = self.innerCircleRadius;
    _dragButton.clipsToBounds = YES;

    UIImage *image = _draggableImage ?: [UIImage imageNamed:@"drag_button"];
    if ([image respondsToSelector:@selector(imageWithRenderingMode:)])
        image = [image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
    [_dragButton setImage:image forState:UIControlStateNormal];

    [self addSubview:_dragButton];

    _dragButton.center = p_fixedCenter(self.frame);
}

#pragma mark - Gesture Recognizers

- (void)addPanGestureRecognizerToImage
{
    UIPanGestureRecognizer *panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(handlePan:)];
    [_dragButton addGestureRecognizer:panGesture];
}

- (void)handlePan:(UIPanGestureRecognizer *)gestureRecognizer
{
    CGPoint _fixedCenter = p_fixedCenter(self.frame);
    if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {
        _distanceMoved = 0.0f;
        if ([_delegate respondsToSelector:@selector(slideOutToUnlockViewDidStartToDrag:)]) {
            [_delegate slideOutToUnlockViewDidStartToDrag:self];
        }
    }

    if (gestureRecognizer.state == UIGestureRecognizerStateChanged) {
        CGPoint translation = [gestureRecognizer translationInView:self];
        _dragButton.center = CGPointMake(_fixedCenter.x + translation.x,
                                         _fixedCenter.y + translation.y);
        _distanceMoved = sqrt(pow(_dragButton.center.x-_fixedCenter.x, 2) + pow(_dragButton.center.y-_fixedCenter.y, 2));

        if ([self.delegate respondsToSelector:@selector(slideOutToUnlockView:didDragDistance:)])
            [self.delegate slideOutToUnlockView:self didDragDistance:_distanceMoved];

        _unlockOnRelease = _distanceMoved > self.outerCircleRadius + self.innerCircleRadius;
    }

    if (gestureRecognizer.state == UIGestureRecognizerStateEnded) {

        if ([self.delegate respondsToSelector:@selector(slideOutToUnlockViewDidEndToDrag:)])
            [self.delegate slideOutToUnlockViewDidEndToDrag:self];

        if (_unlockOnRelease) {
            if ([self.delegate respondsToSelector:@selector(slideOutToUnlockViewDidUnlock:)])
                [self.delegate slideOutToUnlockViewDidUnlock:self];
        } else {
            if ([self.delegate respondsToSelector:@selector(slideOutToUnlockViewDidNotUnlock:)])
                [self.delegate slideOutToUnlockViewDidNotUnlock:self];
        }

        [UIView animateWithDuration:0.2f animations:^{
            _dragButton.center = _fixedCenter;
        }];
    }
}

@end

File: Classes/swift/GBSlideOutToUnlockView.swift
//
//  GBSlideOutToUnlockView.swift
//  GBSlideOutToUnlockView
//
//  Created by Gustavo Barbosa on 6/11/14.
//  Copyright (c) 2014 Gustavo Barbosa. All rights reserved.
//

import UIKit

@objc protocol GBSlideOutToUnlockViewDelegate {
    @optional func slideOutToUnlockViewDidStartToDrag(slideOutToUnlockView: GBSlideOutToUnlockView)
    @optional func slideOutToUnlockViewDidEndToDrag(slideOutToUnlockView: GBSlideOutToUnlockView)
    @optional func slideOutToUnlockViewDidUnlock(slideOutToUnlockView: GBSlideOutToUnlockView)
    @optional func slideOutToUnlockViewDidNotUnlock(slideOutToUnlockView: GBSlideOutToUnlockView)
    @optional func slideOutToUnlockViewDidDragDistance(slideOutToUnlockView: GBSlideOutToUnlockView, draggedDistance: Float)
}

class GBSlideOutToUnlockView: UIView {
    
    weak var delegate: GBSlideOutToUnlockViewDelegate?
    
    var innerCircleRadius: Float = 25.0
    var outerCircleRadius: Float = 50.0
    
    var innerCircleColor: UIColor?
    var outerCircleColor: UIColor?
    
    var draggableButtonBackgroundColor: UIColor?
    var draggableImageTintColor: UIColor?
    var draggableImage: UIImage?
    
    var _dragButton: UIButton!
    var _unlockOnRelease: Bool = false
    var _distanceMoved: Float = 0.0
    
    init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.clearColor()
        self.innerCircleColor = self.tintColor
        self.outerCircleColor = self.tintColor
        self.draggableButtonBackgroundColor = self.tintColor
        self.draggableImageTintColor = UIColor.whiteColor()
    }
    
    override func drawRect(rect: CGRect) {
        drawCircleWithRadius(innerCircleRadius, color: innerCircleColor)
        drawCircleWithRadius(outerCircleRadius, color: outerCircleColor)
        addRedeemImageAtCenter()
        addPanGestureRecognizerToImage()
    }
    
    func drawCircleWithRadius(radius: Float, color: UIColor!) {
        let context: CGContextRef = UIGraphicsGetCurrentContext()
        CGContextSetLineWidth(context, 1.0)
        CGContextSetStrokeColorWithColor(context, color.CGColor)
        let circleArea = CGRectMake(self.center.x - radius, self.center.y - radius, 2 * radius, 2 * radius)
        CGContextAddEllipseInRect(context, circleArea)
        CGContextStrokePath(context)
    }
    
    func addRedeemImageAtCenter() {
        _dragButton = UIButton(frame: CGRect(x: 0, y: 0, width: 2*self.innerCircleRadius, height: 2*self.innerCircleRadius))
        _dragButton.backgroundColor = self.draggableButtonBackgroundColor
        _dragButton.tintColor = self.draggableImageTintColor
        _dragButton.layer.cornerRadius = self.innerCircleRadius
        _dragButton.layer.masksToBounds = true
        
        var templateImage: UIImage?
        if let rawImage = self.draggableImage {
            templateImage = rawImage.imageWithRenderingMode(.AlwaysTemplate)
        } else {
            templateImage = UIImage(named: "drag_button").imageWithRenderingMode(.AlwaysTemplate)
        }
        
        _dragButton.setImage(templateImage!, forState: .Normal)
        
        self.addSubview(_dragButton)
        _dragButton.center = self.center
    }
    
    func addPanGestureRecognizerToImage() {
        let panGeture = UIPanGestureRecognizer(target: self, action: "handlePan:")
        _dragButton.addGestureRecognizer(panGeture)
    }
    
    func handlePan(gestureRecognizer: UIPanGestureRecognizer) {
        if gestureRecognizer.state == .Began {
            _distanceMoved = 0.0
            delegate?.slideOutToUnlockViewDidStartToDrag?(self)
        }
        
        if gestureRecognizer.state == .Changed {
            let translation = gestureRecognizer.translationInView(self)
            _dragButton.center = CGPoint(x: self.center.x + translation.x, y: self.center.y + translation.y)
            
            let a = _dragButton.center.x - self.center.x
            let b = _dragButton.center.y - self.center.y
            _distanceMoved = sqrtf(powf(a, 2.0) + powf(b, 2.0))
            delegate?.slideOutToUnlockViewDidDragDistance?(self, draggedDistance: _distanceMoved)
            
            _unlockOnRelease = _distanceMoved > outerCircleRadius + innerCircleRadius
        }
        
        if gestureRecognizer.state == .Ended {
            delegate?.slideOutToUnlockViewDidEndToDrag?(self)
            
            if _unlockOnRelease {
                delegate?.slideOutToUnlockViewDidUnlock?(self)
            } else {
                delegate?.slideOutToUnlockViewDidNotUnlock?(self)
            }
            
            UIView.animateWithDuration(0.2, animations: {
                self._dragButton.center = self.center
            })
        }
    }
}
